server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderIT.java

```diff
@@ -0,0 +1,339 @@
+        refreshDiskUsage();
+        assertThat(getShardRoutings(dataNode0Id), empty());
+
+        // increase disk size of node 0 to allow just enough room for one shard, and check that it's rebalanced back
+        fileSystemProvider.getTestFileStore(dataNode0Path).setTotalSpace(minShardSize + WATERMARK_BYTES + 1L);
+        refreshDiskUsage();
+        assertThat(getShardRoutings(dataNode0Id), hasSize(1));
+    }
+
+    private Set<ShardRouting> getShardRoutings(String nodeId) {
+        final Set<ShardRouting> shardRoutings = new HashSet<>();
+        for (IndexShardRoutingTable indexShardRoutingTable : client().admin().cluster().prepareState().clear().setRoutingTable(true)
+                .get().getState().getRoutingTable().index("test")) {
+            for (ShardRouting shard : indexShardRoutingTable.shards()) {
+                assertThat(shard.state(), equalTo(ShardRoutingState.STARTED));
+                if (shard.currentNodeId().equals(nodeId)) {
+                    shardRoutings.add(shard);
+                }
+            }
+        }
+        return shardRoutings;
+    }
+
+    /**
+     * Index documents until all the shards are at least WATERMARK_BYTES in size, and return the size of the smallest shard
+     */
+    private long createReasonableSizedShards() throws InterruptedException {
+        while (true) {
+            final IndexRequestBuilder[] indexRequestBuilders = new IndexRequestBuilder[scaledRandomIntBetween(100, 10000)];
+            for (int i = 0; i < indexRequestBuilders.length; i++) {
+                indexRequestBuilders[i] = client().prepareIndex("test").setSource("field", randomAlphaOfLength(10));
+            }
+            indexRandom(true, indexRequestBuilders);
+            forceMerge();
+            refresh();
+
+            final ShardStats[] shardStatses = client().admin().indices().prepareStats("test")
+                    .clear().setStore(true).setTranslog(true).get().getShards();
+            final long[] shardSizes = new long[shardStatses.length];
+            for (ShardStats shardStats : shardStatses) {
+                shardSizes[shardStats.getShardRouting().id()] = shardStats.getStats().getStore().sizeInBytes();
+            }
+
+            final long minShardSize = Arrays.stream(shardSizes).min().orElseThrow(() -> new AssertionError("no shards"));
+            if (minShardSize > WATERMARK_BYTES) {
+                return minShardSize;
+            }
+        }
+    }
+
+    private void refreshDiskUsage() {
+        ((InternalClusterInfoService) internalCluster().getMasterNodeInstance(ClusterInfoService.class)).refresh();
+        // if the nodes were all under the low watermark already (but unbalanced) then a change in the disk usage doesn't trigger a reroute
+        // even though it's now possible to achieve better balance, so we have to do an explicit reroute. TODO fix this?
+        assertAcked(client().admin().cluster().prepareReroute());
```

Comment:
can we somehow only conditionally do this reroute here, i.e.  if all nodes under low watermark?